### DISCUSSION TITLE
Filtering models by their many to many relationship

### DISCUSSION THREAD

**Discussion by farzadsbakhtiar** [fastapi/sqlmodel#587] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from sqlmodel import SQLModel, Relationship, Field, select

def find_users(workspace: Workspace, min_logins: int):
  """ find all users of the workspace with at least min_logins logins """
  statement = select(User).where(User.logins >= min_logins)
                     .where(workspace in User.workspaces) # -- this doesn't work!
  ...


class UserWorkspace(SQLModel, table=True):
  user_id: int = Field(foreign_key='user.id')
  workspace_id: int = Field(foreidgn_key='workspace.id')


class User(SQLModel, table=True):
  id: int ...
  username: str
  logins: int 
  workspaces: list[Workspace] = Relationship(link_model=UserWorkspace)


class Workspace(SQLModel, table=True):
  id: int ...
```

### Description

I can't find a way to filter models by their relationship if it is a x-to-many relationship. The example above tries to find 
all users of a workspace with at least `min_logins`, but there is no way to filter the users by their workspace.

### Operating System

macOS

### SQLModel Version

0.0.8

### Python Version

3.10.0

**Comment by 8thgencore** [comment, marked as answer]
To filter models by their relationship in a x-to-many relationship, you can use the any_ method of the relationship attribute. Here's how you can modify your code to filter users by their workspace:

```python
from sqlmodel import SQLModel, Relationship, Field, select, any_

def find_users(workspace: Workspace, min_logins: int):
  """ find all users of the workspace with at least min_logins logins """
  statement = select(User).where(User.logins >= min_logins) \
                           .where(User.workspaces.any_(id=workspace.id))
  ...

class UserWorkspace(SQLModel, table=True):
  user_id: int = Field(foreign_key='user.id')
  workspace_id: int = Field(foreign_key='workspace.id')

class User(SQLModel, table=True):
  id: int ...
  username: str
  logins: int 
  workspaces: list[Workspace] = Relationship(link_model=UserWorkspace)

class Workspace(SQLModel, table=True):
  id: int ...
```

In this modified code, the any_ method is used to filter the users by their workspace. The id attribute of the Workspace model is used to filter the users, and it is compared to the workspace.id attribute, which is the ID of the workspace passed to the find_users function.

**Reply by farzadsbakhtiar** [reply]
Thank you very much, that solves the problem.
Where is the reference documentation for this?

**Reply by antont** [reply]
[function sqlalchemy.sql.expression.any_](https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.any_)

**Reply by farzadsbakhtiar** [reply]
Is there a doc that describes the relationship of sqlalchemy and sqlmodel?

**Reply by antont** [reply]
All db operations of SQLModel are SQLAlchemy code, the select imported there is actually imported from SQLAlchemy etc. So in simple terms SQLModel is SQLAlchemy. It just integrates Pydantic with SQLAlchemy ORM (declarative_base), so that same Pydantic models then work automatically both for SQLAlchemy and e.g. FastAPI use.

Maybe someone knows a nice doc telling the full pic.

**Reply by vlntsolo** [reply]
There might be a typo, in `any_()` method.  I had to replace it with `any()` without underscore to make it work.
